File: delete_node.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

def push(head, data):
    newn = Node(data)
    newn.next = head
    if head:
        head.prev = newn
    return newn

def delete_node(head, del_node):
    if not head or not del_node:
        return head
    if head == del_node:
        head = del_node.next
    if del_node.next:
        del_node.next.prev = del_node.prev
    if del_node.prev:
        del_node.prev.next = del_node.next
    return head

def print_list(head):
    cur = head
    while cur:
        print(cur.data, end=" ")
        cur = cur.next
    print()

if __name__ == "__main__":
    head = None
    for v in [5,4,3,2,1]:
        head = push(head, v)
    print("Original:", end=" ")
    print_list(head)

    cur = head
    while cur and cur.data != 3:
        cur = cur.next
    head = delete_node(head, cur)
    print("After deleting 3:", end=" ")
    print_list(head)

    head = delete_node(head, head)
    print("After deleting head:", end=" ")
    print_list(head)

File: delete_node.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *prev, *next;
} Node;

Node* push(Node* head, int data) {
    Node* newn = (Node*)malloc(sizeof(Node));
    newn->data = data;
    newn->prev = NULL;
    newn->next = head;
    if (head) head->prev = newn;
    return newn;
}

Node* deleteNode(Node* head, Node* del) {
    if (!head || !del) return head;
    if (head == del) head = del->next;
    if (del->next) del->next->prev = del->prev;
    if (del->prev) del->prev->next = del->next;
    free(del);
    return head;
}

void printList(Node* head) {
    Node* cur = head;
    while (cur) {
        printf("%d ", cur->data);
        cur = cur->next;
    }
    printf("\n");
}

int main() {
    Node* head = NULL;
    head = push(head, 5);
    head = push(head, 4);
    head = push(head, 3);
    head = push(head, 2);
    head = push(head, 1); // list: 1 2 3 4 5

    printf("Original: ");
    printList(head);

    // delete the node containing 3 (find it)
    Node* cur = head;
    while (cur && cur->data != 3) cur = cur->next;
    head = deleteNode(head, cur);

    printf("After deleting 3: ");
    printList(head);

    // delete head
    head = deleteNode(head, head);
    printf("After deleting head: ");
    printList(head);

    return 0;
}
